sound/soc/sof/intel/hda-stream.c

```diff
@@ -88,31 +88,46 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
 /*
  * set up Buffer Descriptor List (BDL) for host memory transfer
  * BDL describes the location of the individual buffers and is little endian.
+ *
+ * The typical BDL entries after set up should looks like this:
+ *
+ * bdl[0] from the init_offset,  n entries before wrap,
+ * the e-1 full entries, and the last e entry.
+ *
+ *			     init_offset
+ *				 |
+ * |------|...........|---------S-------|------|........|--------|-------|
+ *  bdl[n] bdl[n+j]    bdl[e-1] | bdl[0] bdl[1] bdl[m]   bdl[n-2] bdl[n-1]
  */
 int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
 			     struct snd_dma_buffer *dmab,
-			     struct hdac_stream *stream)
+			     struct hdac_stream *stream, u32 init_offset)
 {
 	struct sof_intel_dsp_bdl *bdl;
-	int i, offset, period_bytes, periods;
-	int remain, ioc;
+	unsigned int i, period_bytes, entries;
+	u32 bytes, offset, size = 0;
+	bool ioc, set_ioc, is_wrap = false;
 
 	period_bytes = stream->period_bytes;
 	dev_dbg(sdev->dev, "period_bytes:0x%x\n", period_bytes);
 	if (!period_bytes)
 		period_bytes = stream->bufsize;
 
-	periods = stream->bufsize / period_bytes;
+	entries = stream->bufsize / period_bytes;
 
-	dev_dbg(sdev->dev, "periods:%d\n", periods);
+	/* count the last offset fragment entry */
+	if (stream->bufsize % period_bytes)
+		entries++;
 
-	remain = stream->bufsize % period_bytes;
-	if (remain)
-		periods++;
+	/* count the last index fragment entry */
+	if (init_offset % period_bytes)
+		entries++;
+
+	dev_dbg(sdev->dev, "entries:%d\n", entries);
 
```

Comment:
not quite sold on renaming periods to entries? What was the idea here?

Reply:
I change 'periods' to 'entries' as in Code Loader and Trace cases, we don't have period concept there actually, and this is actually the entries number what we are planning to allocate for BDL entries, it is different with period count, actually.

Reply:
@keyonjie does "num_buffers" make more sense. It is a list of buffer descriptors after all.

Reply:
bdl_entries then...

Reply:
OKay, will change.